glean-core/src/upload/mod.rs

```diff
@@ -94,99 +123,172 @@ impl PingUploadManager {
         queue.clear();
     }
 
-    /// Get the next `PingRequest` in queue.
-    pub fn get_next_ping(&self) -> Option<PingRequest> {
+    /// Gets the next `PingRequest` in queue.
+    ///
+    /// # Return value
+    ///
+    /// `PingUploadTask` - see [`PingUploadTask`](enum.PingUploadTask.html) for more information.
+    pub fn get_next_ping(&self) -> PingUploadTask {
+        if !self.has_processed_pings_dir() {
+            return PingUploadTask::Wait;
+        }
+
         let mut queue = self
             .queue
             .write()
             .expect("Can't write to pending pings queue.");
-        queue.pop_front()
+        match queue.pop_front() {
+            Some(request) => PingUploadTask::Upload(request),
+            None => PingUploadTask::Done,
+        }
     }
 }
 
 #[cfg(test)]
 mod test {
+    use std::time::Duration;
+    use std::thread;
+
+    use serde_json::json;
+
+    use super::*;
     use crate::metrics::PingType;
-    use serde_json::Value as JsonValue;
+    use crate::tests::new_glean;
 
-    let uuid = "c0ffeec0-ffee-c0ff-eec0-ffeec0ffeec0";
-    let url = "http://example.com";
-    let body = json!("{}")
+    const UUID: &str = "c0ffeec0-ffee-c0ff-eec0-ffeec0ffeec0";
+    const URL: &str = "http://example.com";
 
     #[test]
     fn test_doesnt_error_when_there_are_no_pending_pings() {
-        let data_dir = tempfile::tempdir().unwrap();
-        let tmpname = data_dir.path().display().to_string();
-
         // Create a new upload_manager
-        let upload_manager = PingUploadManager::new(&tempname);
-
-        // Try and get the next request
-        let request = upload_manager.get_next_ping();
+        let dir = tempfile::tempdir().unwrap();
+        let tmpname = dir.path().display().to_string();
+        let upload_manager = PingUploadManager::new(&tmpname);
+
+        // HACK!
+        // Wait for processing of pending pings directory to finish.
+        while !upload_manager.has_processed_pings_dir() {
```

Comment:
Is there any way, in Rust, to just wait on [this thread](https://github.com/mozilla/glean/pull/696/files#diff-80512c19c26eacea30d18cd504ce6439R81) to be finished?

Reply:
I looked, but didn't find it :( Maybe @badboy knows?

Reply:
If we store the handle, then we can access it and `.join()` on it (that's what the other test did before refactoring)

Reply:
Yes, but that would required that the `process_pings_dir` live outside of the init function and got called separately.

Since we start the `processed_pings_dir` flag as `false` and change it to `true` once the processing is done, it is kind of required that it get's called on init or immediatelly after, so I thought it didn't really make sense for it to be called outside of init.

Reply:
One possibility would be to return a tuple from this `::new` function like `(PingUploadManager, JoinHandle)` and then we culd join on the test, but I didn't go with that because it looked bad.

Reply:
I'm definitely missing something here, but can't we save the [handle from the builder](https://github.com/mozilla/glean/pull/696/files#diff-80512c19c26eacea30d18cd504ce6439R81) in the `PingUploadTask`, giving it just enough visibility to make useful in tests? @badboy ?

Reply:
We would store it in `PingUploadManager`

Reply:
Do you think I should make that change? I realised I can add the attribute to the `PingUploadManager` struct and then put it under a `#[cfg(test)]`, which would make it only part of this struct for testing. 

One last argument in favour of the current hack, is that is mimics exactly how we plan on calling this function in the future. I could even change it to:
```rust
        while upload_manager.get_next_ping() == PingUploadTask::Wait {
            thread::sleep(Duration::from_millis(10));
        }
```
To make that clearer.

Reply:
Looking at it again, `join` takes the handle by value, so we need to store `Option<JoinHandle>` to make it work.
We only `join` in a test, but I kinda dislike sprinkling `#[cfg(test)]` around.

I think for now the `while !upload_manager.has_processed_pings_dir` is a hack, but an okay-ish one. We can come back to it later of course.

Reply:
> ```rust
>         while upload_manager.get_next_ping() == PingUploadTask::Wait {
>             thread::sleep(Duration::from_millis(10));
>         }
> ```
> 
> To make that clearer.

Using `Wait` makes sense here. I'm fine with this change.

Reply:
> Looking at it again, join takes the handle by value, so we need to store Option<JoinHandle> to make it work.

Yes, I was testing that out and just did exactly this :)